vhost-user-backend/src/bitmap.rs

```diff
@@ -0,0 +1,602 @@
+// SPDX-License-Identifier: Apache-2.0
+
+use std::ops::Index;
+use std::os::fd::{AsRawFd, BorrowedFd};
+use std::sync::atomic::{AtomicU8, Ordering};
+use std::sync::{Arc, RwLock};
+use std::{io, ptr};
+use vm_memory::bitmap::{Bitmap, BitmapSlice, WithBitmapSlice};
+use vm_memory::mmap::NewBitmap;
+use vm_memory::{Address, GuestMemoryRegion};
+
+// Size in bytes of the `VHOST_LOG_PAGE`
+const LOG_PAGE_SIZE: usize = 0x1000;
+// Number of bits grouped together as a basic storage unit ("word") in the bitmap.
+const LOG_WORD_SIZE: usize = u8::BITS as usize;
+
+/// A `Bitmap` with an internal `Bitmap` that can be replaced at runtime
+pub trait BitmapReplace: Bitmap {
+    type InnerBitmap: MemRegionBitmap;
+
+    /// Replace the internal `Bitmap`
+    fn replace(&self, bitmap: Self::InnerBitmap);
+}
+
+/// A bitmap relative to a memory region
+pub trait MemRegionBitmap: Sized {
+    /// Creates a new bitmap relative to `region`, using the `logmem` as
+    /// backing memory for the bitmap
+    fn new<R: GuestMemoryRegion>(region: &R, logmem: MmapLogReg) -> io::Result<Self>;
+}
+
+// NOTE:
+// This impl is a quick and dirty hack to allow the tests to continue using
+// `GuestMemoryMmap<()>`. Sadly this is exposed in the public API, but it should
+// be moved to an internal mock library.
+impl BitmapReplace for () {
+    type InnerBitmap = ();
+
+    // this implementation must not be used if the backend sets `VHOST_USER_PROTOCOL_F_LOG_SHMFD`
+    fn replace(&self, _bitmap: ()) {
+        panic!("The unit bitmap () must not be used if VHOST_USER_PROTOCOL_F_LOG_SHMFD is set");
```

Comment:
This message is a little difficult for me to understand. How do we make sure here that  "The unit bitmap ()" is going to be used with `VHOST_USER_PROTOCOL_F_LOG_SHMFD` set?

Reply:
Well, currently this panic is redundant because `InnerBitmap::new(region, logmem)` (for `()` as ` InnerBitmap`) will fail before calling `region.bitmap().replace(())`. 

But, could be two possible ways to fail here:
- a bug in the fronted sending `SET_LOG_BASE`even if the backend doesn't have the `VHOST_USER_PROTOCOL_F_LOG_SHMFD` protocol feature
- and the backend setting the `VHOST_USER_PROTOCOL_F_LOG_SHMFD` protocol feature and using `GuestMemoryMmap<()>`

Reply:
IMHO a bug in the frontend should not crash the backend, but a backend setting `VHOST_USER_PROTOCOL_F_LOG_SHMFD` and using `GuestMemoryMmap<()>` is acceptable to crash. 

BTW if it is impossible to arrive here also if there is a bug in the frontend, is fine, otherwise I'd avoid the panic.